src/suspendable/core.clj

```diff
@@ -25,8 +25,16 @@
 (defn- stop-missing-components [old-system new-system]
   (component/update-system old-system (missing-keys old-system new-system) component/stop))
 
+(defn- object? [x]
+  (instance? clojure.lang.IObj x))
+
+(defn assoc-meta [component key]
+  (if (object? component)
+    (vary-meta component assoc ::key key)
+    component))
+
 (defn- assoc-component-key [system key]
-  (update-in system [key] vary-meta assoc ::key key))
+  (update-in system [key] assoc-meta key))
```

Comment:
`assoc-meta` and `object?` are perhaps not the most apt names, since `object?` could refer to `java.lang.Object` and `assoc-meta` implies that you can associate any key.

So instead, what about:

```clojure
(defn- maybe-vary-meta [obj f & args]
  (if (instance? obj IObj)
    (apply vary-meta obj f args)
    obj))

(defn- assoc-component-key [system key]
  (update-in system [key] maybe-vary-meta assoc ::key key))
```